Core/Src/Custom/proxSensHandler.c

```diff
@@ -11,13 +11,27 @@
 #include "Custom/GPIOHandler.h"
 #include "Custom/proxSensHandler.h"
 #include <stdint.h>
+#include <stdbool.h>
 
-/*Method responsible for checking and switching proximity sensor state using its trigger input and echo output*/
+
+uint8_t vProxSensMeasure(void);
+bool bIsObstacleInRange(uint8_t);
+
+
+
+/*Function responsible for checking and switching proximity sensor state using its trigger input and echo output*/
 uint8_t vProxSensMeasure(void){
 	static uint8_t iMachineState;
+	bool bTrigState;
+
+	bTrigState = bTrigGetState();
+
+	if(!iMachineState){
+		iMachineState = SENSOR_READY;
+	}
```

Comment:
There is way too many IFs in this code and it surely can be refactored to be way more readable and obvious for the reader what is going on here. Take a look at my suggestion - I know it is wron in terms of the syntax, but you should get the general idea of the refactor

```C
uint8_t vProxSensMeasure(void)
{
    // I wander if it is a good idea to define this iMachineState inside of this function
    // I think this static or maybe even non static variable should be created somewhere in the
    // higher scope and passed to this function as a parameter, and should not be modified inside of this
    // function but rather return a brand new value IN THE NAME OF IMMUTABILITY
	static uint8_t iMachineState;
	iMachineState = initializeMachineState(iMachineState);
	
	bool trigState = bTrigGetState();
	bool echoState = bEchoGetState();
	
	switch(iMachineState)
	{
	    case SENSOR_READY:
	    iMachineState = handleSensorReadyState(iMachineState, trigState);
	    break;
	    case SENSOR_STARTING:
	    iMachineState = handleSensorStartingState(iMachineState, echoState, trigState);
	    break;
	    case SENSOR_MEASURE_IN_PROGRESS:
	    iMachineState = handleSensorMeasureInProgressState(iMachineState, echoState);
	    break;
	    case SENSOR_MEASURE_FINNISHED:
	    iMachineState = handleSensorFinishedState(iMachineState);
	    return;
	    case SENSOR_ERROR:
	    iMachineState = handleSensorErrorState(iMachineState);
	    return;
	}

	return iMachineState;
}

uint8_t initializeMachineState(uint8_t iMachineState) // Param should be a ref/pointer to the value - IDK how to I'm a JS dev...
{
    if (!iMachineState)
    {
		return SENSOR_READY;
	}
	return iMachineState;
}

void ensureMachineStateCorrectness(uint8_t iMachineState, uint8_t expectedState)
{
    if(iMachineState != expectedState)
    {
        // Somehow throw error, errno = 1 Operation not permitted
        // Throwing an error should stop the program or do something
        // to break to stop the process or start the fallback logic
    }
}

uint8_t handleSensorReadyState(uint8_t iMachineState, bool trigState)
{
    ensureMachineStateCorrectness(iMachineState, SENSOR_READY);
    if(trigState)
    {
		vTrigChangeState();
		return SENSOR_STARTING;
	}
	return SENSOR_ERROR;
}

uint8_t handleSensorStartingState(uint8_t iMachineState, bool echoState, bool trigState)
{
    ensureMachineStateCorrectness(iMachineState, SENSOR_STARTING);
    if (echoState && trigState)
    {
		vTrigChangeState();
		return SENSOR_MEASURE_IN_PROGRESS;
	}
	return iMachineState;
}

uint8_t handleSensorMeasureInProgressState(uint8_t iMachineState, bool echoState)
{
    ensureMachineStateCorrectness(iMachineState, SENSOR_MEASURE_IN_PROGRESS);
    return echoState ? SENSOR_MEASURE_IN_PROGRESS : SENSOR_MEASURE_FINNISHED
}

uint8_t handleSensorFinishedState(uint8_t iMachineState)
{
    ensureMachineStateCorrectness(iMachineState, SENSOR_MEASURE_FINNISHED);
    return SENSOR_READY;
}

uint8_t handleSensorErrorState(uint8_t iMachineState)
{
    ensureMachineStateCorrectness(iMachineState, SENSOR_ERROR);
    return SENSOR_READY;
}
```

Reply:
Its very likely that this sensor is too fast for project developed using FreeRTOS, at least on this microcontroller - one tick of FreeRTOS takes 1ms, and there are parts of measuring process taking no more than couple μs. I will most likely change it to other sensor whitch will make this code unnecessary, it will be taken care of along with senor swap